src/main/java/com/xebialabs/overthere/smb/SmbConnection.java

```diff
@@ -0,0 +1,211 @@
+        port = addressPort.getPort();
+
+        int unmappedSmbPort = options.getInteger(SMB_PORT, PORT_DEFAULT_SMB);
+        InetSocketAddress smbAddressPort = mapper.map(createUnresolved(unmappedAddress, unmappedSmbPort));
+        smbPort = smbAddressPort.getPort();
+        username = options.get(USERNAME);
+        password = options.get(PASSWORD);
+        client = new SMBClient(new DefaultConfig());
+    }
+
+    public void connect() {
+        createConnection();
+        connected();
+    }
+
+    private void createConnection() {
+        try {
+            UserAndDomain ud = getUserNameAndDomain(username);
+            String user = ud.getUsername();
+            String domain = ud.getDomain();
+            connection = client.connect(hostname);
+            AuthenticationContext authContext = new AuthenticationContext(user, password.toCharArray(), domain);
+            session = connection.authenticate(authContext);
+        } catch (IOException e) {
+            throw new RuntimeIOException(e);
+        } catch (NtlmException e) {
+            throw new RuntimeIOException(e);
+        }
+    }
+
+    @Override
+    public OverthereFile getFile(String hostPath) {
+        hostPath = SmbPaths.escapeForwardSlashes(hostPath);
+        Map<String, String> pathMappings = options.get(PATH_SHARE_MAPPINGS, PATH_SHARE_MAPPINGS_DEFAULT);
+        return new SmbFile(this, hostPath, pathMappings);
+    }
+
+    @Override
+    public OverthereFile getFile(OverthereFile parent, String child) {
+        return parent.getFile(child);
+    }
+
+    @Override
+    protected void doClose() {
+        try {
+            for (DiskShare s : shareCache.values())
+                try {
+                    s.close();
+                } catch (IOException e) {
+                    logger.warn("Exception while trying to close smb share", e);
+                }
+        } finally {
+            shareCache.clear();
+            try {
+                if (session != null) {
+                    session.close();
+                }
+            } catch (IOException e) {
+                logger.warn("Exception while trying to close smb session", e);
+            } finally {
+                try {
+                    if (connection != null) {
+                        connection.close();
+                    }
+                } catch (Exception e) {
+                    logger.warn("Exception while trying to close smb connection", e);
+                }
+            }
+        }
+    }
+
+    @Override
+    protected OverthereFile getFileForTempFile(OverthereFile parent, String name) {
+        return getFile(parent, name);
+    }
+
+    @Override
+    public String toString() {
+        return "smb:" + cifsConnectionType.toString().toLowerCase() + "://" + username + "@" + hostname + ":" + smbPort + ":" + port;
+    }
+
+    protected DiskShare getShare(String shareName) {
+        DiskShare share = shareCache.get(shareName);
```

Comment:
I'll add the cache in SMBJ, as it should be there (see section 3.2.4.2 of the MS-SMB2 spec, the TreeConnectTable).

It will be there in 0.0.6